### PR TITLE
Add-SearchPickupRequest

### PR DESCRIPTION
**Pull request by louis1706**
Sorry I used Prefix 

Adding the SearchPickup Event

EventArgs has:
Player
Pickup
SearchTime
IsAllowed

### CODE REVIEW

File: Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    using Mirror.LiteNetLib4Mirror;
+
+    /// <summary>
+    /// Patches <see cref="SearchCoordinator.ReceiveRequestUnsafe"/>.
+    /// Adds the <see cref="Handlers.Player.SearchPickupRequesting"/> event.
+    /// </summary>
+    [HarmonyPatch(typeof(SearchCoordinator), nameof(SearchCoordinator.ReceiveRequestUnsafe))]
+    internal static class SearchPickupRequesting
+    {
+        private static void Prefix(SearchCoordinator __instance, ref bool __result, out SearchSession? session, out SearchCompletor completor)
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs:30]: Return bool value here

File: Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            try
+            {
+                SearchRequest request = __instance.SessionPipe.Request;
+                completor = SearchCompletor.FromPickup(__instance, request.Target, __instance.ServerMaxRayDistanceSqr);
+                SearchPickupRequestEventArgs ev = new SearchPickupRequestEventArgs(Player.Get(__instance.Hub), request.Target, request.Target.SearchTime, completor.ValidateStart());
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs:36]: Pass to the constructor all the arguments you can pass, and make the constructor able to accept them

File: Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                }
+
+                session = new SearchSession?(body);
+                __result = true;
+                return;
+            }
+            catch (Exception exception)
+            {
+                Log.Error($"{typeof(SearchPickupRequesting).FullName}.{nameof(Prefix)}:\n{exception}");
+                session = null;
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs:70]: No

File: Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                completor = null;
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs:71]: Same as above

File: Exiled.Events/Handlers/Player.cs [hunk 5]
```diff
@@ -422,6 +422,12 @@ public static class Player
         /// </summary>
         public static event CustomEventHandler<InteractingScp330EventArgs> InteractingScp330;
 
+        /// <summary>
+        /// Called before pre-authenticating a player.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:426]: Fix docs

File: Exiled.Events/Handlers/Player.cs [hunk 6]
```diff
@@ -902,5 +908,11 @@ public static class Player
         /// </summary>
         /// <param name="ev">The <see cref="InteractingScp330EventArgs"/> instance.</param>
         public static void OnInteractingScp330(InteractingScp330EventArgs ev) => InteractingScp330.InvokeSafely(ev);
+
+        /// <summary>
+        /// Called before a <see cref="Exiled.API.Features.Player"/> interacts with an Pickup.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:913]: Same as above

File: Exiled.Events/Handlers/Player.cs [hunk 7]
```diff
@@ -422,6 +422,12 @@ public static class Player
         /// </summary>
         public static event CustomEventHandler<InteractingScp330EventArgs> InteractingScp330;
 
+        /// <summary>
+        /// Called before pre-authenticating a player.
+        /// Invoked before a <see cref="Exiled.API.Features.Player"/> interacts with Pickup.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:427]: Same as above

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    using System;
+
+    using Exiled.API.Enums;
+    using Exiled.API.Features;
+    using Exiled.API.Features.Items;
+
+    using InventorySystem.Items.Pickups;
+
+    /// <summary>
+    /// Contains all information after the end of a round.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs:19]: Fix docs

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        public SearchPickupRequestEventArgs(Player player, ItemPickupBase pickup, float searchTime, bool isAllowed)
+        {
+            Player = player;
+            Pickup = Pickup.Get(pickup);
+            SearchTime = searchTime;
+            IsAllowed = isAllowed;
+        }
+
+        /// <summary>
+        /// Gets the Player.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs:39]: Same as above

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        public Player Player { get; }
+
+        /// <summary>
+        /// Gets the searching Pickup.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs:44]: Same as above

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        public Pickup Pickup { get; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the Pickup can be search.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs:49]: Same as above

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        public float SearchTime { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the Pickup can be search.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs:54]: Same as above

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]
Rename the event to `SearchingPickupEventArgs`.
Fix all docs.
Improve the event and the patch.

File: Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs (hunk 2 above)
Comment (louis1706) [reply Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs]: i have added the two other value

File: Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs (hunk 3 above)
Comment (louis1706) [reply Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs]: i need to put this for don't getting Error when i compiled

File: Exiled.Events/Handlers/Player.cs (hunk 5 above)
Comment (louis1706) [reply Exiled.Events/Handlers/Player.cs]: Fix

File: Exiled.Events/Handlers/Player.cs (hunk 7 above)
Comment (louis1706) [reply Exiled.Events/Handlers/Player.cs]: Fix

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs (hunk 9 above)
Comment (louis1706) [reply Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs]: i have added some things not sure if it's correct for you

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs (hunk 10 above)
Comment (louis1706) [reply Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs]: i have added some things not sure if it's correct for you

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs (hunk 8 above)
Comment (louis1706) [reply Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs]: I have changed it

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs (hunk 12 above)
Comment (louis1706) [reply Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs]: Idk by what i can replace this docs

File: Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs (hunk 11 above)
Comment (louis1706) [reply Exiled.Events/EventArgs/SearchPickupRequestEventArgs.cs]: this docs it's fix

File: Exiled.Events/Handlers/Player.cs (hunk 6 above)
Comment (louis1706) [reply Exiled.Events/Handlers/Player.cs:912]: /// Called before a <see cref="Exiled.API.Features.Player"/> search an Pickup.
is it good ?

File: Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs (hunk 1 above)
Comment (louis1706) [reply Exiled.Events/Patches/Events/Player/SearchPickupRequesting.cs]: it's done i return the bool value for make the base method not call when the Prefix have any error

File: Exiled.Events/EventArgs/SearchingPickupEventArgs.cs [hunk 13]
```diff
@@ -14,29 +14,33 @@ namespace Exiled.Events.EventArgs
     using Exiled.API.Features.Items;
 
     using InventorySystem.Items.Pickups;
+    using InventorySystem.Searching;
 
     /// <summary>
-    /// Contains all information after the end of a round.
+    /// Contains all information before an player search an pickup.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchingPickupEventArgs.cs:20]: Fix grammar

File: Exiled.Events/EventArgs/SearchingPickupEventArgs.cs [hunk 14]
```diff
@@ -14,29 +14,33 @@ namespace Exiled.Events.EventArgs
...
+            SearchSession = searchSession;
+            SearchCompletor = searchCompletor;
             SearchTime = searchTime;
-            IsAllowed = isAllowed;
+            IsAllowed = searchCompletor.ValidateStart();
         }
 
         /// <summary>
-        /// Gets the Player.
+        /// Gets the Player who's getting the Pickup.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchingPickupEventArgs.cs:43]: Same as above

File: Exiled.Events/EventArgs/SearchingPickupEventArgs.cs [hunk 15]
```diff
@@ -46,7 +50,17 @@ public SearchPickupRequestEventArgs(Player player, ItemPickupBase pickup, float
         public Pickup Pickup { get; }
 
         /// <summary>
-        /// Gets or sets a value indicating whether the Pickup can be search.
+        /// Gets or sets the SearchSession value.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchingPickupEventArgs.cs:53]: Fix docs style

File: Exiled.Events/EventArgs/SearchingPickupEventArgs.cs [hunk 16]
```diff
@@ -46,7 +50,17 @@ public SearchPickupRequestEventArgs(Player player, ItemPickupBase pickup, float
...
+        /// </summary>
+        public SearchSession SearchSession { get; set; }
+
+        /// <summary>
+        /// Gets or sets the SearchCompletor value.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchingPickupEventArgs.cs:58]: Fix docs style

File: Exiled.Events/EventArgs/SearchingPickupEventArgs.cs [hunk 17]
```diff
@@ -46,7 +50,17 @@ public SearchPickupRequestEventArgs(Player player, ItemPickupBase pickup, float
...
+        /// </summary>
+        public SearchCompletor SearchCompletor { get; set; }
+
+        /// <summary>
+        /// Gets or sets the duration time for getting the Pickup.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/SearchingPickupEventArgs.cs:63]: Fix grammar

File: Exiled.Events/Handlers/Player.cs [hunk 18]
```diff
@@ -423,10 +423,9 @@ public static class Player
         public static event CustomEventHandler<InteractingScp330EventArgs> InteractingScp330;
 
         /// <summary>
-        /// Called before pre-authenticating a player.
-        /// Invoked before a <see cref="Exiled.API.Features.Player"/> interacts with Pickup.
+        /// Invoked before a <see cref="Exiled.API.Features.Player"/> Search a Pickup.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Player.cs:426]: Fix grammar & docs style

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

**Comment by Thundermaker300** [comment]
Think I got the grammar fixes :)

**Review by Michal78900 — APPROVED** [review]

**Review by NaoUnderscore — APPROVED** [review]

**Pull request merged**